modules/module_service.go

```diff
@@ -40,6 +40,13 @@ func NewModuleService(name string, logger log.Logger, service services.Service,
 	}
 
 	w.Service = services.NewBasicService(w.start, w.run, w.stop)
+
+	if namedService, isNamed := service.(services.NamedService); isNamed {
+		return services.NamedServiceDecorator{
+			Service: w,
+			Name:    namedService.ServiceName(),
```

Comment:
I don't think we should cache `namedService.ServiceName()` into a field, but call it each time. Service may update its name dynamically (eg. to include state or some other value).

Reply:
> Service may update its name dynamically (eg. to include state or some other value).

This is actually quite common in Guava services.

Reply:
we should probably not cache them in mimir either.. https://github.com/grafana/mimir/blob/7aa6015704d6b5c080f5fd66dff9303df122563e/pkg/mimir/mimir.go#L438

but got your point - will refactor this

Reply:
> we should probably not cache them in mimir either.. https://github.com/grafana/mimir/blob/7aa6015704d6b5c080f5fd66dff9303df122563e/pkg/mimir/mimir.go#L438

I'm not sure I understand what you're saying. I don't see any place in Mimir where we would call `ServiceName` and cache it. In fact the only place where it is used is in failure watcher (indirectly, via `DescribeService`).

Reply:
you're right. what I linked is caching module names. seem like I'm prone to confusing the two